clients/src/main/java/org/apache/kafka/common/metrics/JmxReporter.java

```diff
@@ -26,99 +25,161 @@
+    private final Registry registry;
+    private final MBeanServer mbeanServer;
 
+    /**
+     * Create a JMX reporter.
+     */
     public JmxReporter() {
         this("");
     }
 
     /**
      * Create a JMX reporter that prefixes all metrics with the given string.
+     *
+     * @param prefix        The string to prefix all metrics with.
      */
     public JmxReporter(String prefix) {
+        this(prefix, Registry.INSTANCE, ManagementFactory.getPlatformMBeanServer());
+    }
+
+    /**
+     * Create a JMX reporter that prefixes all metrics with the given string
+     * and uses the given MBeanServer.
+     *
+     * @param prefix        The string to prefix all metrics with.
+     * @param registry      The registry that stores beans.
+     * @param mbeanServer   The MBean server to use for JMX operations.
+     */
+    public JmxReporter(String prefix, Registry registry, MBeanServer mbeanServer) {
         this.prefix = prefix;
+        this.registry = registry;
+        this.mbeanServer = mbeanServer;
     }
 
     @Override
     public void configure(Map<String, ?> configs) {}
 
     @Override
     public void init(List<KafkaMetric> metrics) {
-        synchronized (LOCK) {
-            for (KafkaMetric metric : metrics)
-                addAttribute(metric);
-            for (KafkaMbean mbean : mbeans.values())
-                reregister(mbean);
-        }
-    }
-
-    public boolean containsMbean(String mbeanName) {
-        return mbeans.containsKey(mbeanName);
+        addMetrics(metrics);
     }
 
     @Override
     public void metricChange(KafkaMetric metric) {
-        synchronized (LOCK) {
-            KafkaMbean mbean = addAttribute(metric);
-            reregister(mbean);
-        }
+        addMetrics(Collections.singleton(metric));
     }
 
     @Override
     public void metricRemoval(KafkaMetric metric) {
-        synchronized (LOCK) {
-            MetricName metricName = metric.metricName();
-            String mBeanName = getMBeanName(prefix, metricName);
-            KafkaMbean mbean = removeAttribute(metric, mBeanName);
-            if (mbean != null) {
-                if (mbean.metrics.isEmpty()) {
-                    unregister(mbean);
-                    mbeans.remove(mBeanName);
-                } else
-                    reregister(mbean);
+        removeMetrics(Collections.singleton(metric));
+    }
+
+    // Visible for testing
+    public boolean containsMbean(String mbeanName) {
+        return registry.findMBeanOwner(mbeanName) == this;
+    }
+
+    void addMetrics(Collection<KafkaMetric> metrics) {
+        Map<String, List<KafkaMetric>> beanNameToMetrics = groupMetrics(metrics);
+        for (Map.Entry<String, List<KafkaMetric>> entry : beanNameToMetrics.entrySet()) {
+            KafkaMbean mbean = registry.getOrCreateLockedMBean(entry.getKey(), this);
+            if (mbean == null) {
+                log.warn("Bean name conflict: {} is already registered to a different " +
```

Comment:
Could we show the name of JmxReporter which has the conflict bean name? It is useful to trace this rare issue.

for another, Is throwing exception more suitable than logging?

Reply:
The JmxReporter doesn't really have a useful name.  It would just print a memory address.

Throwing an exception would be a behavior change which I don't think is a good idea here-- too much chance of breaking something.

Reply:
> The JmxReporter doesn't really have a useful name. It would just print a memory address.

How about showing the prefix of JmxReporter? It helps us to trace the "different" JmxReporter.

Reply:
> Throwing an exception would be a behavior change which I don't think is a good idea here-- too much chance of breaking something.

IIRC, previous behavior is to "override" the previous beans so this PR has introduced the breaking change.

Reply:
The prefix of the JmxReporter will show up in the bean name, so it's not necessary to print again.

The previous behavior is not as simple as "overriding existing beans."  In particular, whenever attributes are created or deleted ownership could flip-flop between MetricsReporter objects.

Reply:
Thanks for explanation!